### PR TITLE
document #15618

### PR DESCRIPTION
**Pull request by ringabout**

### CODE REVIEW

**Comment by timotheecour** [comment]
Can we fix it instead?

File: lib/pure/unittest.nim [hunk 1]
```diff
@@ -92,6 +92,15 @@
...
 ##
 ##     echo "suite teardown: run once after the tests"
+## 
+## Limitations/Bugs
+## ================
+## Since `check` will rewrite some programs for
+## supporting checkpoints(Namely assigns expressions to variables).
+## Some type conversions are not supported.
+## For example `check 4.0 == 2 + 2` won't work. But `doAssert 4.0 == 2 + 2` works.
+## Make sure the same type on both sides of the operator(such as `=`, `==`, `>=` and so on).
```
Comment (ringabout) [inline lib/pure/unittest.nim]:
```suggestion
## Make sure the same type on both sides of the operator(such as `==`, `>=` and so on).
```

**Comment by ringabout** [comment]
I think we should make this work to fix this issue.
```nim
var x = 4
doAssert 4.0 == x
```

**Comment by timotheecour** [comment]
> I think we should make this work to fix this issue.

no, this can't work.
But `check 4.0 == 2 + 2` should work

**Comment by ringabout** [comment]
`check 4.0 == 2 + 2` is converted to this for supporting checkpoints. It seems hard to fix.
```nim
var x = 2 + 2    # maybe have side effects
doAsssert 4.0 == x
checkpointes(... & $x)
```

And

```nim
func double*(n: float): float = n * 2
doAssert(double(-2) == -4)
test "double -2": check(double(-2) == -4)
````

is converted to:

```nim
func double*(n: float): float = n * 2

var x = double(-2) # maybe have side effects
doAssert(x == -4)
checkpointes(... & $x)
```

**Comment by timotheecour** [comment]
> It seems hard to fix.

check lhs == rhs
=>
var rhs2: type(lhs) = rhs


note that in future work `{.evalonce.}` might be worth considering eg: `let rhs2 {.evalonce.}: type(lhs) = expr` refs https://github.com/nim-lang/Nim/pull/13750

**Comment by ringabout** [comment]
Thanks for hints! I will give a try.

But how about this?
```nim
import unittest


check 2.0 + 2.0 == 2 + 2  # works with doAssert
check 2 + 2 == 2.0 + 2.0 # works with doAssert

var x: int = 2.0 + 2.0 # This failed.
```

or lhs and rhs are both functions.

It may not be hard. But I think It is error-prone.

**Comment by timotheecour** [comment]
`nnkIntLit` is special, its type is determined from context.

here's a working prototype:

```nim
when true:
  import macros
  macro commonType(a, b: typed): untyped =
    if a.kind == nnkIntLit:
      result = quote do:
        type(`b`)
    else:
      result = quote do:
        type(`a`)

  macro check2(cond: untyped): untyped =
    let op=cond[0]
    let a1=cond[1]
    let a2=cond[2]
    let s = cond.repr.newLit
    result = quote do:
      type T = commonType(`a1`, `a2`)
      let lhs: T = `a1`
      let rhs: T = `a2`
      if not(`op`(lhs, rhs)):
        doAssert false, $(lhs, rhs, `s`)
  check2(2+2 == 2.0 + 2.0)
  check2(2.0 + 2.0 == 2+2)
  check2(2+3 == 2.0 + 2.0) # this will show the errmsg
```

File: lib/pure/unittest.nim [hunk 2]
```diff
@@ -92,6 +92,15 @@
 ##         discard v[4]
 ##
 ##     echo "suite teardown: run once after the tests"
+## 
+## Limitations/Bugs
+## ================
+## Since `check` will rewrite some programs for
+## supporting checkpoints(Namely assigns expressions to variables).
```
Comment (juancarlospaco) [inline lib/pure/unittest.nim]:
```suggestion
## supporting checkpoints (namely assigns expressions to variables).
```

File: lib/pure/unittest.nim [hunk 3]
```diff
@@ -92,6 +92,15 @@
...
 ##
 ##     echo "suite teardown: run once after the tests"
+## 
+## Limitations/Bugs
+## ================
+## Since `check` will rewrite some programs for
+## supporting checkpoints (namely assigns expressions to variables).
+## Some type conversions are not supported.
+## For example `check 4.0 == 2 + 2` won't work. But `doAssert 4.0 == 2 + 2` works.
+## Make sure the same type on both sides of the operator(such as `==`, `>=` and so on).
```
Comment (Araq) [inline lib/pure/unittest.nim]:
```suggestion
## Make sure both sides of the operator (such as `==`, `>=` and so on) have the same type.
```

**Pull request merged**